test-suite/src/test/groovy/io/micronaut/EventListenerSpec.groovy

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut
+
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.annotation.Requires
+import io.micronaut.context.event.ApplicationEvent
+import io.micronaut.context.event.ApplicationEventListener
+import io.micronaut.context.event.ShutdownEvent
+import io.micronaut.context.event.StartupEvent
+import io.micronaut.core.order.OrderUtil
+import io.micronaut.discovery.event.ServiceReadyEvent
+import io.micronaut.discovery.event.ServiceStoppedEvent
+import io.micronaut.inject.qualifiers.Qualifiers
+import io.micronaut.runtime.event.annotation.EventListener
+import io.micronaut.runtime.server.EmbeddedServer
+import io.micronaut.runtime.server.event.ServerShutdownEvent
+import io.micronaut.runtime.server.event.ServerStartupEvent
+import jakarta.inject.Singleton
+import spock.lang.RepeatUntilFailure
+import spock.lang.Specification
+import spock.util.concurrent.PollingConditions
+
+class EventListenerSpec extends Specification {
+
+    @RepeatUntilFailure(maxAttempts = 100) // Retry the test multiple times to test the listeners order
+    void "test all events listener is invoked"() {
+        given:
+        PollingConditions conditions = new PollingConditions(timeout: 10)
+        EmbeddedServer embeddedServer = ApplicationContext.run(EmbeddedServer,
+                ['spec.name': getClass().getSimpleName(),
+                'micronaut.application.name': 'events-test']
+        )
+
+        when:
+        TestAllEventsListener t = embeddedServer.getApplicationContext().getBean(TestAllEventsListener)
+        def serverStartupListeners = embeddedServer.getApplicationContext().getBeansOfType(ApplicationEventListener.class, Qualifiers.byTypeArguments(ServerStartupEvent))
+                    .stream()
+                    .sorted(OrderUtil.COMPARATOR)
+                    .toArray(ApplicationEventListener[]::new)
+        def serverShutdownListeners = embeddedServer.getApplicationContext().getBeansOfType(ApplicationEventListener.class, Qualifiers.byTypeArguments(ServerShutdownEvent))
+                    .stream()
+                    .sorted(OrderUtil.COMPARATOR)
+                    .toArray(ApplicationEventListener[]::new)
+        then:
+        serverStartupListeners.last().getClass().name.contains "NettyServiceDiscovery"
+            serverShutdownListeners.last().getClass().name.contains "NettyServiceDiscovery"
+
+        conditions.eventually {
+            t.events.size() == 3
+        }
+
+        embeddedServer.close()
+
+        conditions.eventually {
+            t.events.size() == 6
+        }
+
+        t.events[0] == StartupEvent
+        t.events[1] == ServerStartupEvent
+        t.events[2] == ServiceReadyEvent
+        t.events[3] == ServerShutdownEvent
+        t.events[4] == ServiceStoppedEvent
```

Comment:
should not the order be 

```
StartupEvent
ServerStartupEvent
ServiceReadyEvent
ServiceStoppedEvent
ServerShutdownEvent
ShutdownEvent
```

instead of

```
StartupEvent
ServerStartupEvent
ServiceReadyEvent
ServerShutdownEvent
ServiceStoppedEvent
ShutdownEvent
```
?

Reply:
The service events are published by the `NettyServiceDiscovery` reacting on the server started /  shutdown

Reply:
the event order is correct in the PR